proposals/0000-forced-class-instantiation.rst

```diff
@@ -0,0 +1,142 @@
+.. proposal-number::
+.. trac-ticket::
+.. implemented::
+.. highlight:: haskell
+
+This is a GHC proposal `under discussion`_.
+
+.. _`under discussion`: https://github.com/ghc-proposals/ghc-proposals/pull/23
+
+Forced Class Instantiation
+==========================
+
+A declaration
+::
+  instance force Cls Ty
+
+instantiates any ``Cls a`` constraint encountered, including in error messages, otherwise ambiguous types.
+
+Motivation
+----------
+
+I am teaching a Haskell class right now, and in the first few lectures, I had to tell them: „Whenever you see ``Foldable t => … t a …`` anywhere, just mentally replace this with ``… [a] …``.”. Similarly, when discussing ``IO`` before discussing `Monad`, I had to tell them the same for these two. Finally, the first few weeks (because of using CodeWorld_), we turned on ``OverloadedStrings``, but I really do not want want to see them `IsString a` and rather have them see ``Text`` in error messages.
+
+.. _CodeWorld: http://code.world/haskell
+
+The students coped, but it would be better if the compiler did that for them. So the main motivation is to tell them to include a line like
+::
+  instance force Foldable []
+
+in their file (students accept instructions to include stuff they do not understand yet, so this is fine) and get a lower entry level.
+
+But it is useful not only for teaching. Especially
+::
+  instance IsString Text
+is useful for anyone who does not want literals overloaded but really wants them to be monomorphically the desired ``Text`` type.
+
+Proposed Change
+---------------
+I propose to introduce a declaration 
+::
+  instance force Cls Ty1 Ty2 …
+  
+The declaration is valid if the mentioned instance exists (i.e. if the constraing ``Cls Ty1 Ty2`` can be completely solved). If not, as in ``instance force Eq [a]``, a usual error message (“Could not solve constraint Eq [a]: No instance [a]” or similar) is produced, and the program is rejected.
+
+The declaration has the effect that, within the scope of one module, the existance of the type class ``Cls`` is hidden from the user, by consistenly instantiationg it with the instance given in the declaration.
```

Comment:
I have no clue what this sentence means, and that's important because it's the nub of the specification.  Precisely when do you "consistently instantiate it", and what does that even mean?

I think, but I am not sure, that your attention is focused SOLELY on the moment at which we infer the type for a let-bound function. (If every function had a type signature, then there would be no effect at all.)

So perhaps you intend that GHC should not generalise over a constraint that matches one of these force-instance things?   

But presumably if you had  `force instance C [Int]` and you had a constraint `C [alpha]` you would not instantly apply the force-instance in case we discovered that `alpha ~ Bool`?  Or are you not allowed to solve `C [Boo]`?

What if there are two force-instance things; does it matter which you default first?